### PR TITLE
[orchagent] TWAMP Light orchagent implementation

### PR DESCRIPTION
**Pull request by huseratgithub**
**What I did**
 Implemented the TWAMP Light feature according to the SONiC TWAMP Light HLD(https://github.com/sonic-net/SONiC/pull/1320).

**Why I did it**
Part of the TWAMP Light feature support.

**How I verified it**
It Implemented VS test.

**Details if related**
sonic-swss-common: https://github.com/sonic-net/sonic-swss-common/pull/823
sonic-sairedis: https://github.com/sonic-net/sonic-sairedis/pull/1306

### CODE REVIEW

**Comment by prsunny** [comment]
@AlanYoush , please rebase and resolve conflicts

**Comment by huseratgithub** [comment]
@prsunny PR has been rebased and all conflicts have been resolved.
Could you please help to review & merge, thanks.

**Comment by huseratgithub** [comment]
@prsunny kindly reminder.
Could you please review or assign someone to review the PR, thanks

**Comment by AlanYoush** [comment]
@prsunny
This PR is very important for TWAMP Light. And PR (https://github.com/sonic-net/sonic-sairedis/pull/1306) has been merged.  We expect TWAMP Light to be pushed in the release version of 202405. 
So please help to review this PR or invite others to review it. And If you can merge it, please help to merge.
Thanks very much.

**Comment by AlanYoush** [comment]
@prsunny just a kind reminder

**Comment by prsunny** [comment]
> **What I did** Implemented the TWAMP Light feature according to the SONiC TWAMP Light HLD(#1320).
> 
> **Why I did it** Part of the TWAMP Light feature support.
> 
> **How I verified it** It Implemented VS test.
> 
> **Details if related** sonic-swss-common: [sonic-net/sonic-swss-common#823](https://github.com/sonic-net/sonic-swss-common/pull/823) sonic-sairedis: [sonic-net/sonic-sairedis#1306](https://github.com/sonic-net/sonic-sairedis/pull/1306)

Can you please fix the link for HLD?

**Comment by prsunny** [comment]
@huseratgithub , as mentioned in HLD PR, i cannot find a sonic-mgmt PR for this feature? Do you plan to provide one?

File: orchagent/saihelper.cpp [hunk 1]
```diff
@@ -82,6 +82,7 @@ sai_dash_inbound_routing_api_t*     sai_dash_inbound_routing_api;
 sai_dash_eni_api_t*                 sai_dash_eni_api;
 sai_dash_vip_api_t*                 sai_dash_vip_api;
 sai_dash_direction_lookup_api_t*    sai_dash_direction_lookup_api;
+sai_twamp_api_t*            sai_twamp_api;
```
Comment (prsunny) [inline orchagent/saihelper.cpp:85]: Please align this with the previous code.

File: orchagent/twamporch.cpp [hunk 2]
```diff
@@ -0,0 +1,1083 @@
...
+void TwampOrch::removeSessionStatus(const string& name)
+{
+	SWSS_LOG_ENTER();
+
+	m_stateDbTwampTable.del(name);
+}
+
+void TwampOrch::removeSessionCounter(const sai_object_id_t session_id)
+{
+	SWSS_LOG_ENTER();
```
Comment (prsunny) [inline orchagent/twamporch.cpp:281]: Alignment wrong

File: orchagent/twamporch.cpp [hunk 3]
```diff
@@ -0,0 +1,1083 @@
...
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Unable to query the TWAMP event notification capability");
+        return false;
+    }
+
+    if (!capability.set_implemented)
+    {
+        SWSS_LOG_ERROR("TWAMP register event notification not supported");
```
Comment (prsunny) [inline orchagent/twamporch.cpp:337]: IMO, this is not an error but a notice as SAI may not implement this feature. Please change

File: orchagent/twamporch.cpp [hunk 4]
```diff
@@ -0,0 +1,1083 @@
...
+    sai_status_t status;
+    sai_attr_capability_t capability;
+
+    status = sai_query_attribute_capability(gSwitchId, SAI_OBJECT_TYPE_SWITCH,
+                                            SAI_SWITCH_ATTR_TWAMP_SESSION_EVENT_NOTIFY,
+                                            &capability);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Unable to query the TWAMP event notification capability");
```
Comment (prsunny) [inline orchagent/twamporch.cpp:331]: IMO, this is not an error but a notice as SAI may not implement this feature. Please change

File: orchagent/twamporch.cpp [hunk 5]
```diff
@@ -0,0 +1,1083 @@
...
+    {
+        m_maxTwampSessionCount = attr.value.s32;
+    }
+
+    /* Set MAX entries to counter DB */
+    vector<FieldValueTuple> fvTuple;
+    fvTuple.emplace_back("MAX_TWAMP_SESSION_COUNT", to_string(m_maxTwampSessionCount));
+    m_switchOrch->set_switch_capability(fvTuple);
+
+    /* Add TWAMP session event notification support */
```
Comment (prsunny) [inline orchagent/twamporch.cpp:142]: Why do we need to register this if SAI capability is returned as false? Please skip this section if SAI returned false for switch capability

**Review by prsunny — CHANGES_REQUESTED** [review]

**Comment by huseratgithub** [comment]
> @huseratgithub , as mentioned in HLD PR, i cannot find a sonic-mgmt PR for this feature? Do you plan to provide one?

@prsunny This is the initial commit for phase 1. I plan to do sonic-mgmt in later phase.

File: orchagent/saihelper.cpp (hunk 1 above)
Comment (huseratgithub) [reply orchagent/saihelper.cpp]: Done

File: orchagent/twamporch.cpp (hunk 2 above)
Comment (huseratgithub) [reply orchagent/twamporch.cpp]: Done

File: orchagent/twamporch.cpp (hunk 4 above)
Comment (huseratgithub) [reply orchagent/twamporch.cpp]: Done

File: orchagent/twamporch.cpp (hunk 5 above)
Comment (huseratgithub) [reply orchagent/twamporch.cpp:145]: Done

**Comment by huseratgithub** [comment]
Now all tests passed

File: orchagent/twamporch.cpp (hunk 3 above)
Comment (huseratgithub) [reply orchagent/twamporch.cpp]: Done

**Comment by huseratgithub** [comment]
> Please rebase to latest master and fix the coverage issue.

Hi @prsunny , The code coverage check passed.

**Comment by huseratgithub** [comment]
@prsunny Many thanks for the review. All comments have been addressed.
Could you please help to approve this PR ? Thanks.

**Comment by huseratgithub** [comment]
@prsunny just a kind reminder.
This is for [SONiC 202405 Release](https://github.com/orgs/sonic-net/projects/17) and also CLI is ready, so it is better to complete before branch out.
Please help to approve or to assign someone on behalf of you.
Thank you very much.

**Comment by huseratgithub** [comment]
@prsunny just a kind reminder

**Comment by huseratgithub** [comment]
@prsunny just a kind reminder

File: orchagent/twamporch.cpp (hunk 5 above)
Comment (prsunny) [reply orchagent/twamporch.cpp:145]: I didn't get the change here. My comment is to skip the entire registration process if SAI capability is returned as false. Where is the capability check?

Comment (huseratgithub) [reply orchagent/twamporch.cpp:142]: Need to check capability when creating entry, the default value 0 means no resources and return failed(see line 512, please).
Also I will add skip if the SAI capability is returned as false.
Thank you again for your comment.

Comment (huseratgithub) [reply orchagent/twamporch.cpp:147]: Done, all tests passed. Please review again, thank you very much.

File: orchagent/twamporch.cpp [hunk 6]
```diff
@@ -0,0 +1,1052 @@
...
+    m_maxTwampSessionCount = m_twampSessionCount = 0;
+
+    /* Get the Maximum supported TWAMP sessions */
+    SWSS_LOG_INFO("Get the Maximum supported TWAMP sessions");
+    sai_attribute_t attr;
+    attr.id = SAI_SWITCH_ATTR_MAX_TWAMP_SESSION;
+    sai_status_t status = sai_switch_api->get_switch_attribute(gSwitchId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_WARN("Twamp session resource availability is not supported. Skipping ...");
```
Comment (prsunny) [inline orchagent/twamporch.cpp:126]: Please change to NOTICE as this is expected to be not supported on some platforms

File: orchagent/twamporch.cpp [hunk 7]
```diff
@@ -0,0 +1,1052 @@
...
+    /* Set MAX entries to state DB */
+    if (m_maxTwampSessionCount)
+    {
+        vector<FieldValueTuple> fvTuple;
+        fvTuple.emplace_back("MAX_TWAMP_SESSION_COUNT", to_string(m_maxTwampSessionCount));
+        m_switchOrch->set_switch_capability(fvTuple);
+    }
+    else
+    {
+        SWSS_LOG_WARN("Twamp session resource availability is not supported. Skipping ...");
```
Comment (prsunny) [inline orchagent/twamporch.cpp:143]: Please change to NOTICE as this is expected to be not supported on some platforms

Comment (huseratgithub) [reply orchagent/twamporch.cpp]: Done.

File: orchagent/twamporch.cpp (hunk 6 above)
Comment (huseratgithub) [reply orchagent/twamporch.cpp]: Done

**Comment by prsunny** [comment]
Can you please rebase?

**Comment by huseratgithub** [comment]
> Can you please rebase?

Hi @prsunny , thank you for the reminder.
I have rebased to latest now, please review and approve.

**Comment by liushilongbuaa** [comment]
@huseratgithub , @prsunny ,Please help with PR validation failure.
https://github.com/sonic-net/sonic-buildimage/pull/18334

**Comment by huseratgithub** [comment]
> @huseratgithub , @prsunny ,Please help with PR validation failure. [sonic-net/sonic-buildimage#18334](https://github.com/sonic-net/sonic-buildimage/pull/18334)

Thank you for the reminder. It looks like 'hw_stats' uninitialized.

**Pull request merged**